src/meshes/t8code_mesh.jl

```diff
@@ -157,27 +197,49 @@ function T8codeMesh(trees_per_dimension; polydeg,
 
     # Get cell length in reference mesh: Omega_ref = [-1,1]^2.
-    dx = 2 / trees_per_dimension[1]
-    dy = 2 / trees_per_dimension[2]
+    dx = [2 / n for n in trees_per_dimension]
 
     num_local_trees = t8_cmesh_get_num_local_trees(cmesh)
 
     # Non-periodic boundaries.
     boundary_names = fill(Symbol("---"), 2 * NDIMS, prod(trees_per_dimension))
 
+    if mapping === nothing
+        mapping_ = coordinates2mapping(ntuple(_ -> -1.0, NDIMS), ntuple(_ -> 1.0, NDIMS))
+    else
+        mapping_ = mapping
+    end
+
     for itree in 1:num_local_trees
         veptr = t8_cmesh_get_tree_vertices(cmesh, itree - 1)
         verts = unsafe_wrap(Array, veptr, (3, 1 << NDIMS))
 
-        # Calculate node coordinates of reference mesh.
-        cell_x_offset = (verts[1, 1] - 1 / 2 * (trees_per_dimension[1] - 1)) * dx
-        cell_y_offset = (verts[2, 1] - 1 / 2 * (trees_per_dimension[2] - 1)) * dy
+        if NDIMS == 2
+            # Calculate node coordinates of reference mesh.
+            cell_x_offset = (verts[1, 1] - 0.5 * (trees_per_dimension[1] - 1)) * dx[1]
+            cell_y_offset = (verts[2, 1] - 0.5 * (trees_per_dimension[2] - 1)) * dx[2]
 
-        for j in eachindex(nodes), i in eachindex(nodes)
-            tree_node_coordinates[:, i, j, itree] .= mapping(cell_x_offset +
-                                                             dx * nodes[i] / 2,
-                                                             cell_y_offset +
-                                                             dy * nodes[j] / 2)
+            for j in eachindex(nodes), i in eachindex(nodes)
+                tree_node_coordinates[:, i, j, itree] .= mapping_(cell_x_offset +
+                                                                  dx[1] * nodes[i] / 2,
+                                                                  cell_y_offset +
+                                                                  dx[2] * nodes[j] / 2)
+            end
+        elseif NDIMS == 3
+            cell_x_offset = (verts[1, 1] - 0.5 * (trees_per_dimension[1] - 1)) * dx[1]
+            cell_y_offset = (verts[2, 1] - 0.5 * (trees_per_dimension[2] - 1)) * dx[2]
+            cell_z_offset = (verts[3, 1] - 0.5 * (trees_per_dimension[3] - 1)) * dx[3]
+
+            for k in eachindex(nodes), j in eachindex(nodes), i in eachindex(nodes)
+                tree_node_coordinates[:, i, j, k, itree] .= mapping_(cell_x_offset +
+                                                                     dx[1] * nodes[i] / 2,
+                                                                     cell_y_offset +
+                                                                     dx[2] * nodes[j] / 2,
+                                                                     cell_z_offset +
+                                                                     dx[3] * nodes[k] / 2)
+            end
+        else
+            throw(ArgumentError("NDIMS should be 2 or 3."))
```

Comment:
I guess you could simplify this by calling `@assert NDIMS == 2 || NDIMS == 3` in line 213 which spares you from the last else.